[1 - 2단계 - 방탈출 예약 대기] 부기(이창의) 미션 제출합니다.

changuii opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?

#### Repository
기존의 repository를 JPA로 전환하는 과정에서, repository 계층을 제거하고, 바로 JpaRepository를 상속받은 인터페이스로 전환하였습니다.

때문에, 기존에 사용하던 fake 객체를 사용하지 못했고 새로 테스트 모두 작성하게 되었습니다.
- 또한, JpaRepository에 정의되어 있는 메서드들이 매우 많기 때문에 이를 구현한 Fake를 직접 구현할 수 없어, Mockito를 이용하여 모든 메서드들을 Mocking하거나 Stubbing하여 구현하였습니다.

이를 통해 fake를 따로 관리해주어야 하는 번거로움은 사라졌지만, 변경하고 생각해보니 코드를 다시 JDBC로 바꿀 수는 없는 코드가 되어버렸습니다.
- JpaRepository를 하나의 계층으로 바라보기 보다는, JDBC Template처럼, 데이터베이스에 접근하기 위한 repository로 바라보고, 실제로 repository 계층을 앞에 두는게 추후에, 데이터베이스 접근 방식의 변경에 더 쉽게 대응할 수 있겠다.라는 생각이 들었습니다.

다시,, 계층을 하나더 추가해야할까요? 
현재 : controller -> service -> JpaRepository
변경 : controller -> service -> repository -> JpaRepository

#### 집중적으로 보고 싶은 부분들

api url을 한번 검토받고 싶습니다!

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

changuii committed:
init: 프로젝트 가져오기

changuii committed:
refactor: ThemeRepository를 JpaRepository로 변경 및 FakeThemeRepository 삭제

- ThemeRepository를 Spring Data JPA 기반으로 변경하여 의존성을 단순화하고 성능 및 유지보수성을 개선.
- FakeThemeRepository 및 관련 테스트 코드 삭제.
- ThemeJdbcRepository 및 관련 초기 데이터(.sql) 삭제.

changuii committed:
refactor: RDB 기반의 repository를 JPA 기반으로 전환

기존 JdbcTemplate을 활용하던 repository 구현체를 제거하고, JPA repository로 리팩토링하여 코드 단순성과 유지보수성을 개선하였습니다. 이에 따라 관련 도메인 엔티티 클래스에 JPA 매핑 어노테이션(@Entity 등)을 추가하였으며, 연관 관계를 매핑하여 관리하도록 수정하였습니다. 추가적으로 사용되지 않는 schema.sql 제거 및 예약 관련 유효성 검사 로직을 간소화하였습니다.

changuii committed:
refactor: @AllArgsConstructor를 활용하여 생성자 제거 및 불필요한 @Autowired 어노테이션 삭제

- Lombok의 @AllArgsConstructor를 추가하여 생성자를 자동 생성하도록 변경
- @Autowired 어노테이션 삭제 및 관련 수동 생성자 제거
- 클래스 필드의 final 제거, 가변성을 고려한 수정
- 코드 가독성과 유지보수성을 향상시킴

changuii committed:
test: Mock 기반으로 Fake Repository 제거 및 테스트 리팩토링

changuii committed:
feat: reservation-mine 페이지를 응답하라

changuii committed:
test: reservation-mine 페이지를 응답하라

changuii committed:
feat: 예약 상태를 나타내는 ReservationStatus 추가 및 멤버별 예약 조회 기능 구현

- `ReservationStatus` enum 추가 및 예약 상태 관리
- `MemberReservationResponse` 클래스 추가로 예약 조회 DTO 정의
- 멤버별 예약 조회 엔드포인트(`/members/reservations`) 및 서비스 로직 추가
- `Member` 엔티티에 `reservations` 연관관계 설정
- SQL 스키마에 `reservation_status` 필드 추가 및 초기 데이터 수정

changuii committed:
test: 멤버의 예약 조회 테스트 추가

Hyunta commented on src/main/java/roomescape/auth/AuthController.java:
```diff
@@ -14,19 +14,13 @@
 import roomescape.auth.dto.LoginResponse;
 
 @RestController
+@AllArgsConstructor
```
`@AllArgsConstructor` 는 어떤 역할을 하나요?

changuii replied:
##### 역할과 동작
AllArgsConstructor는 모든 필드를 가진, 생성자를 구현해줍니다. 
실제로 롬복은 compileOnly 의존성으로 compile time classpath에 존재하기 때문에 실제 실행 시점에는 어노테이션이 존재하지 않고, 컴파일을 하면서 롬복이 해당 클래스 내부에 생성자를 만들어줍니다!
- lombok을 적용함으로써 반복적으로 나오는 코드들을 줄일 수 있었습니다. 마치 record를 사용하는 것과 비슷하다고 생각합니다!

##### 장점과 사용 이유
AuthController에 이렇게 모든 필드를 초기화하는 생성자를 만들어준 이유는 필드 변수들에 대한 생성자 주입을 사용하기 위함입니다.
필드 주입을 사용할 수도 있지만 저는 다음과 같은 이유로 생성자 주입을 사용하였습니다.

생성자 주입을 사용한 이유는 필드를 final 키워드를 사용하여 인스턴스 레퍼런스가 변경되지 않도록 만들어 실수를 방지하고, 명시적으로 이 필드의 레퍼런스는 바뀌지 않는다를 전달하고 싶었습니다. 
- 추가로 객체가 생성된 시점에 모든 필드가 초기화되어 있음을 확실히할 수도 있습니다. 덕분에 필드가 아직 초기화되지 않은 것은 아닐까? 라는 고민을 할 필요가 없어집니다.

##### 단점
단점은 다음과 같습니다!
1. lombok이라는 라이브러리에 대해 프로젝트의 의존도가 높아진다. 만약 lombok이 지원을 중단하게 된다면 문제가 될 수 있지도 않을까? 라는 생각이 들긴 했습니다. 다만, 오픈소스로써, 오랜 기간 지원해주고 있기 때문에 큰 문제가 되지 않을 것이라 생각하였습니다.
2. 프로젝트가 커질수록 빌드에 걸리는 시간이 증가함, 빌드를 수행하면서 어노테이션을 제거하고, 어노테이션에 해당하는 코드를 생성해주어야 하기 때문
3. 컴파일 타임에 코드가 존재하지 않고 어노테이션으로 존재하기 때문에 디버깅이 어려울 수 있습니다. 다만, 현재 롬복으로 생성된 코드에서는 디버깅할만한 코드가 있는지는 잘 모르겠습니다.

##### 결론
저는 lombok을 믿고 반복적으로 작성해야하는 코드들을 롬복 어노테이션을 통해 줄이고, 다른 개발자가 코드를 읽을 때 불필요한 정보들을 감출 수 있는 이점을 활용하기 위해 lombok을 사용할 것입니다.
이는 앞에서 설명한 사용할때의 단점 보다는 장점이 더 크다고 생각했습니다.

Hyunta commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -1,48 +1,60 @@
+
+@Entity
+@AllArgsConstructor
+@EqualsAndHashCode(onlyExplicitlyIncluded = true)
```
이렇게 작성하면 어떤 방식의 EqualsAndHashCode를 만들어주나요?

changuii replied:
이렇게 작성하면 명시적으로 `@EqualsAndHashCode.Include` 어노테이션이 붙은 필드만을 사용하여 equals()와 hashCode()를 만들어 줍니다.
of 속성에 문자열로 값을 넣어줄 수도 있지만 해당 방법은 문자열로 필드 명을 작성하는 방식이기 때문에 리팩터링을 하거나 실수로 잘못된 값을 넣을 가능성이 존재합니다. 따라서 저는 현재 방식을 사용하여 구현하였습니다! 
추가로 엔티티는 id를 기준으로 같다, 아니다를 판단하기 때문에 id필드만을 사용하였습니다!

Hyunta replied:
어떻게 구현되는지도 한번 확인해보시면 좋을 것 같습니다

https://projectlombok.org/features/EqualsAndHashCode

Hyunta commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -1,48 +1,60 @@
 public class Member {
 
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
```
`@GeneratedValue`는 어떤 역할을 하나요? 다른 전략은 어떤게 있을까요?

changuii replied:
##### 각 전략별 장단점과 동작
1. auto 전략은 데이터베이스마다 지원하는 id 생성 전략을 선택해서 사용할 수 있도록 해주지만, 네이티브 쿼리를 사용하고 있다면 주의해야 한다. 즉, 결국 조금이라도 의존하고 있다.

2. indentity 전략은 auto_increment 전략을 사용하기 때문에 데이터베이스에 반드시 한 번 저장을 해야만 id를 생성할 수 있다. 따라서 성능 최적화가 힘들다. 따라서 쓰기 지연을 사용할 수 없게 되고 반드시 flush가 발생하게 된다.

3. sequence 전략은 sequence 함수를 사용하기 때문에 데이터베이스에서 시퀀스 함수 조회를 통해 id를 생성한다. identity와 다른 점은 시퀀스 값은 미리 캐싱해두고 사용할 수 있기 때문에 최적화할 수 있는 요소가 많다.

4. table 전략은 id 생성용 테이블을 만들어서 사용하는 전략이기 때문에 데이터베이스와 관계없이 사용할 수 있다. 하지만, 동시성 문제에 취약할 수 있다.

5. ddl 전략은 id jpa가 uuid를 생성해서 사용하기 때문에 데이터베이스에 의존되지 않고, 동시성 문제도 거의 발생하지 않으면서(충돌 가능성 매우 낮음), 쓰기 지연도 사용가능하다. 하지만 uuid의 길이가 길기 때문에 생성하는데 오버헤드가 있으며, id를 기준으로 정렬하기가 까다로울 수 있다. 다른 전략들과는 다르게, 테이블 ddl을 생성에도 추가되는 옵션은 없다. 

##### 내 생각
데이터베이스의 변경이 자주 발생하는 환경이 아니라면 굳이 auto와 table 전략을 고려할 필요는 없다고 생각이 됩니다.
- auto는 native query를 사용하지 못한다는 단점이 존재
- table은 동시성 문제를 고려해야하는 비용과 테이블이 많아진다는 단점이 존재

만약 id를 기준으로 정렬을 할 필요가 없고 분산 환경이라면 UUID를 고려
간단한 애플리케이션을 만들고 auto_increment를 사용할 수 있다면 identity 전략을 고려
많은 데이터 insert가 존재하고 알아보기 쉬운 id를 사용하고 싶다면 sequence 전략을 고려

이번 미션에서는 많은 양의 insert가 필요하지는 않고, 간단한 애플리케이션이기 때문에 auto_increment 전략을 선택할 것 같습니다.

Hyunta commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -1,48 +1,60 @@
+    private String password;
+    private String name;
+
+    @Enumerated(EnumType.STRING)
```
Enumerated는 어떤 애노테이션인가요?

changuii replied:
enum을 테이블에 매핑하기 위한 어노테이션입니다.
##### 옵션별 장단점
1. STRING: enum 객체의 이름을 사용하여 데이터베이스에 저장합니다. 따라서, enum 객체의 순서에 영향을 받지 않지만 enum 객체의 이름을 변경한다면 문제가 발생할 수 있습니다.
2. ORDINAL: enum 객체의 순서 번호를 사용하여 데이터베이스에 저장합니다. 따라서 enum 객체의 이름 변경에는 영향을 받지 않지만 enum 객체의 순서가 변경된다면 문제가 발생할 수 있습니다.
3. enum을 class로 변경하고 다대일 관계 매핑: 앞에서 말한 문제는 발생하지 않지만, 조회시 JOIN 쿼리가 필요하기 때문에 성능상으로 비교적 느립니다.

##### 생각
개인적인 생각으로는 enum 객체의 이름보다는 객체의 순서가 바뀔 가능성이 더 크고, 객체의 순서는 인지를 못할 가능성이 존재한다고 생각합니다.
추가로, enum 객체의 순서를 변경했을 때 데이터베이스도 이에 대응하기 위해 변경해주는 상황을 상상해봤을 때, 

###### ORDINAL
ORDINAL을 사용한다면, enum 객체의 순서가 변경되었을 경우 최소 2개의 객체가 변경되었을 것이고 따라서 수정 작업을 두 번에 걸쳐 수정해야합니다. 이때 실수가 발생한다면 매우 큰 문제가 발생할 수 있습니다.
###### STRING
하지만, STRING을 사용한다면, enum 객체의 이름이 변경되었을 경우 하나의 객체만 변경해주면 되기 때문에 비교적 쉽고 숫자의 경우 실수로 다른 번호로 중복되게 변경한다면 돌이킬 수 없지만, 객체의 이름의 경우에는 문자열이기 때문에 다른 객체의 이름으로 실수로 수정할 가능성이 비교적 작다고 생각이 듭니다.

###### 결론 
따라서, 저는 enum 객체의 이름이 enum 객체의 순서보다 변하지 않을 것이라고 생각합니다. 따라서 이름을 신중하게 정해서, enum 객체의 순서에 상관없이 사용할 수 있도록 STRING 옵션을 사용하였습니다.

Hyunta commented on src/main/java/roomescape/member/Member.java:
```diff
@@ -1,48 +1,60 @@
+    @Enumerated(EnumType.STRING)
+    private MemberRole role;
+
+    @OneToMany(mappedBy = "member")
```
`@OneToMany`는 어떤 역할을 하나요? `@ManyToOne`과 어떻게 다른가요?

changuii replied:
먼저 `@ManyToOne`은, 현재 엔티티가 나타내는 테이블이 상대 테이블과 다대일 관계를 가진다는 매핑 정보를 의미합니다.
즉, 데이터베이스 관점에서 바라보았을 때, 현재 엔티티는 상대 엔티티를 1개를 가지고, 상대 엔티티는 현재 엔티티를 여러개를 가진다는 의미입니다.
따라서, 외래키가 현재 테이블에서 관리하고 있다는 의미가 됩니다.

`@OneToMany`는 방금 설명한 매핑 관계와 반대되는 개념입니다. 현재 테이블 엔티티 하나에 상대 테이블 엔티티를 여러개 가진다는 의미입니다.
따라서 외래키는 상대 테이블에 있습니다.
하지만, 실제 외래키는 상대방에게 있기 때문에 mappedBy를 통해 누가 연관관계의 주인인지를 설정해줄 수 있습니다.
- 이 연관관계의 주인은 필드의 변경이 데이터베이스로 반영된다는 의미이고, 연관관계의 주인이 아닌 필드는 쓰기가 불가능한 읽기 전용으로 사용되어지기 때문에 필드의 변경이 데이터베이스에 반영되지는 않습니다.

현재 멤버가 가지는 reservation들을 조회하는 경우가 존재하기 때문에 `OneToMany`를 사용하여 주었고 읽기 전용으로 만들어주었습니다.

Hyunta commented on src/main/java/roomescape/theme/ThemeRepository.java:
```diff
@@ -2,16 +2,19 @@
-
-    Theme findById(Long id);
-    List<Theme> findAll();
+public interface ThemeRepository extends JpaRepository<Theme, Long> {
+    @Query(nativeQuery = true,
+            value = """
+                            SELECT t.id, t.name, t.description, t.thumbnail
+                            FROM theme t INNER JOIN reservation r
+                                ON t.id=r.theme_id
+                            WHERE r.date BETWEEN ?1 AND ?2
+                            GROUP BY t.id, t.name, t.description, t.thumbnail
+                            ORDER BY COUNT(*) DESC
+                            LIMIT ?3
+                    """)
```
해당 쿼리를 개선해볼 수 있을까요?

changuii replied:
##### 개선 전 쿼리
```java
SELECT t.id, t.name, t.description, t.thumbnail
FROM theme t INNER JOIN reservation r
	ON t.id=r.theme_id
WHERE r.date BETWEEN ?1 AND ?2
GROUP BY t.id, t.name, t.description, t.thumbnail
ORDER BY COUNT(*) DESC
LIMIT ?3
```
현재 쿼리에서 개선할 방법을 고민해보았습니다.

1. WHERE절에 작성한 내용은 reservation에 대한 필터링만 거치고 있기 때문에 굳이, 조인 후에 하는 것보다는 reservation에서 먼저 필터링 한 후, 조인한다면 reservation 컬럼 수가 줄어, 조인 성능이 높아질 것이라 판단하였습니다.
2. GROUP BY도 theme_id에 대한 예약의 수를 카운팅해야하기 때문에, 먼저 group을 맺어, 카운트해준다면 reservation의 수를 더 줄일 수 있다고 판단하였습니다.
3. group by 절에 그룹핑할 컬럼의 수도 줄이면 좋겠다고 판단하여 꼭 필요한 theme_id만 묶도록 하였습니다.

##### 개선한 쿼리 

```java
SELECT t.id, t.name, t.description, t.thumbnail
FROM theme t INNER JOIN (
	    SELECT r.theme_id AS theme_id, COUNT(*) AS count
	    FROM reservation r
	    WHERE r.date BETWEEN ?1 AND ?2
	    GROUP BY r.theme_id
    ) AS r
ON t.id=r.theme_id
ORDER BY r.count DESC
LIMIT ?3
```
그래서 FROM 절에서 서브 쿼리를 통해 먼저 필터링과 그룹핑 작업을 해준 후, 조인을 하여 개선하였습니다!

Hyunta replied:
잘 개선해주셨네요, 추가로 `@Param` 을 사용하면 `?1` 처럼 사용하지 않고 변수명을 직접 명시해줄 수 있습니다.

Hyunta commented on src/main/resources/application.properties:
```diff
@@ -1,3 +1,8 @@
+spring.datasource.url=jdbc:h2:mem:database;
+spring.jpa.show-sql=true
+spring.jpa.properties.hibernate.format_sql=true
+spring.jpa.ddl-auto=create-drop
```
해당 옵션은 어떤 기능을 하나요?

changuii replied:
- `spring.datasource.url`은 사용할 데이터베이스 연결 기술, 데이터베이스 드라이버, 위치,  데이터베이스 스키마를 나타냅니다. 만약 jdbc, mysql, 127.0.0.1, 3306 port, mydatabase scheme를 사용한다면 다음과 같이 작성할 수 있습니다. `jdbc:mysql://127.0.0.1:3306/mydatabase`
- `spring.jpa.show-sql`은 jpa가 sql 구문을 데이터베이스로 전송할 때, 어떤 쿼리를 보내는지 확인할 수 있는 옵션입니다.
- `spring.jpa.properties.hibernate.format_sql=true`은, 데이터베이스 sql을 보기 좋게 포맷팅하여 출력할지 여부를 true, false로 입력받는 옵션입니다!
- `spring.jpa.ddl-auto`은 매핑 정보를 바탕으로 ddl을 사용하는 옵션을 지정합니다!
	- `none`: ddl로 테이블을 생성해주는 작업 없음
	- `validate`: ddl 매핑 정보들을 바탕으로 실제 데이터베이스와 매핑 정보가 일치하는지를 검증한다. 만약 다르다면 애플리케이션이 종료된다.
	- `update`: ddl 매핑 정보들을 바탕으로 다른 부분에 대한 ddl을 작성하여 생성한다.
	- `create`: ddl 매핑 정보들을 바탕으로 시작할 때, ddl을 생성한다.
	- `create-drop`: ddl 매핑 정보들을 바탕으로 시작할 때, ddl을 생성하고 종료할 때, 테이블들을 삭제한다.

Hyunta commented on src/main/java/roomescape/theme/ThemeRepository.java:
```diff
@@ -2,16 +2,19 @@
 
```
> 다시,, 계층을 하나더 추가해야할까요?

JPA로 변환을 하면서 인터페이스에 의존을 하는 구조를 만들다보니 기존에 사용하던 코드들에서 변경해야할 부분들이 많았을 것 같습니다. 말씀해주신 것처럼 추상화가 되어있다면 쉽게 갈아낄 수 있는 구조가 됩니다. 그런데 인터페이스에 필요한 쿼리의 메서드를 추가하는 것 만으로는 모자라집니다. 추가된 계층에도 메서드를 가지고 있어야지 사용할 수 있으니까요. Spring Data JPA를 사용하면서 누리는 장점을 하나 잃어버리게 되는 셈이죠.

DB의 접근하는 기술은 그렇게 자주 바뀌는 부분은 아니라고 생각됩니다. 빈도를 생각해봤을 떄 이 경우에는 확장성 보다는 편리하게 사용할 수 있도록 계층을 추가하지 않는 방향이 적절하지 않을까요? 어떻게 생각하시나요?

changuii replied:
아서가 말씀하신 것처럼, 제가 생각했을 때도 DB 접근 방식은 일반적으로 자주 변경되는 요소는 아니지만,  
**추상화 계층을 생략해서 얻는 단순함보다, 추상화를 도입해 확장성을 확보하는 것이 장기적으로 더 큰 이점**을 가진다고 생각합니다.

특히, 내가 책임지고 유지보수해야 할 코드라면  
향후 DB 접근 기술이 변경되거나 테스트가 필요할 때  
**유연하게 대처할 수 있는 구조를 미리 만들어두는 것이 더 책임감 있는 접근**이라고 판단합니다.

단기적인 편리함보다 **변화에 대한 대비와 코드에 대한 주인의식**이 더 중요하다고 생각합니다.

Hyunta reviewed (changes requested):
안녕하세요 부기,
방탈출 예약 대기 미션 리뷰어로 만나게 된 아서입니다.
미션을 잘 진행해주셨네요, 관련해서 몇가지 커멘트 남겼습니다.
확인해보시고 궁금하거나 모호한 부분이 있으면 DM 주세요

changuii committed:
refactor: 랭킹 조회 쿼리 성능 최적화 및 썸네일 URL 수정

랭킹 조회 SQL 쿼리를 서브쿼리 형태로 변경하여 성능을 개선하였으며, init_data.sql의 테마 썸네일 URL을 Wikimedia 이미지 링크로 변경.

changuii committed:
refactor: application.properties 파일을 application.yml로 변경 및 설정 구조 리팩토링

changuii committed:
refactor: 인증 관련 클래스 위치 변경

인증 관련 클래스(AuthenticationPrincipal, AuthenticationPrincipalArgumentResolver 등)를 `auth` 패키지에서 `config` 패키지로 이동하여 패키지 구조 정리 및 코드 가독성을 개선했습니다.

changuii committed:
feat: PasswordEncoder 추가 및 회원 등록 시 비밀번호 암호화 적용

- PasswordEncoder 컴포넌트를 추가하여 비밀번호 암호화 및 매칭 기능 구현
- MemberService에서 회원 등록 시 비밀번호를 암호화하도록 수정
- init_data.sql에 암호화된 비밀번호 데이터 반영(password는 1234)

changuii committed:
test: PasswordEncoder 관련 테스트 추가 및 기존 테스트에 적용

PasswordEncoder 테스트를 추가하고, AdminViewTest, MemberServiceTest, AuthServiceTest 등 기존 테스트에서 PasswordEncoder를 사용하도록 수정하여 비밀번호 검증 로직을 반영함.

changuii reviewed (commented):
말씀해주신 부분을 제외하고 다음의 내용을 추가하였습니다.! 
- 비밀번호 암호화 추가

changuii committed:
feat: @Column 및 제약 조건 추가

- Member, ReservationTime, Reservation, Theme 엔티티에 @Column 어노테이션 추가
- 각 필드에 unique, nullable 등의 제약 조건 명시
- Reservation 엔티티에 reservationStatus 속성 필수값으로 설정
- 관련 테스트 코드 INSERT 문 수정하여 reservation_status 값 추가

Hyunta reviewed (approved):
안녕하세요 부기,
남겨주신 커멘트에 대한 답변 달아드렸습니다.
궁금하시거나 모호한 부분이 있으면 DM 주세요.
이번 단계는 여기서 머지하도록 하겠습니다.
